### PR TITLE
[Backfill] flink based upsert table backfill support

### PR DESCRIPTION
**Pull request by rohityadav1993**
`feature` `release-notes` `backward-incompatible`

Follow up of #13107

Delivers: https://github.com/apache/pinot/issues/12987, https://github.com/apache/pinot/issues/10896

This PR adds support for uploading batch generated segments to an upsert table. A new segmentName convention `UploadedRealtimeSegmentName` was introduced which enables segment assignment for uploaded segments(not generated by stream consumers) based on the provided partition number.

Changes:
- Fixes an edge case bug with segment flush happening on flushThreshold + 1
- Adds new config in BatchConfigMap needed for `UploadedRealtimeSegmentNameGenerator`
- Flink PinotSinkFunction uses `UploadedRealtimeSegmentNameGenerator` for generating segments for Upsert tables for correct segment assignment.
- Adds support to configure segment name prefix in Flink PinotSinkFunction
- Adds support to configure segment creation time for segments generated as part of a run of Flink workload. This is helpful to group segments of a run with the same date and if needed refresh/overwrite using a rerun.
- Adds a new integration test for upsert tables
- Refactor README with an example to quick start.

Tagged backward-incompatible as it changes the constructor signature for FlinkSegmentWriter

Test plan:
- Integration test
    - Test partitioned parallel segment creation and uplaod
    - Test partitioned parallel segment creation with increasing sequenceId
- Cluster test:
  - Set up a new upsert table with keys partitioned using murmur2
  - Generate segments from a Flink job and upload to table
  - Validation:
    - Check the keys of a uploaded segment are consistently partitioned with realtime stream generated segments
    - Screenshot 2024-08-19 at 2 26 49 PM

### CODE REVIEW

**Comment by rohityadav1993** [comment]
@klsince, could you help with the review given you have context of the prior PR https://github.com/apache/pinot/pull/13107.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 1]
```diff
@@ -60,13 +63,17 @@ public class PinotSinkFunction<T> extends RichSinkFunction<T> implements Checkpo
   private TableConfig _tableConfig;
   private Schema _schema;
 
+  private String _segmentNamePrefix;
+  @Nullable private Long _overriddenSegmentCreationTimeMs;
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:67]: let's keep the class variable name as `segmentCreationTimeMs` here as having `overridden` may be confusing given we are not overriding any other value in this class with this variable.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 2]
```diff
@@ -49,6 +50,8 @@ public class PinotSinkFunction<T> extends RichSinkFunction<T> implements Checkpo
   public static final int DEFAULT_EXECUTOR_POOL_SIZE = 5;
   public static final long DEFAULT_EXECUTOR_SHUTDOWN_WAIT_MS = 3000;
 
+  public static final String DEFAULT_UPLOADED_REALTIME_SEGMENT_PREFIX = "flink";
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:53]: should we keep the default prefix as "uploaded_flink" for more clarity?

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 3]
```diff
@@ -124,12 +132,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
...
     String segmentSuffix = batchConfigMap.get(segmentNamePostfixProp);
     segmentSuffix = segmentSuffix == null ? String.valueOf(_indexOfSubtask) : segmentSuffix + "_" + _indexOfSubtask;
     batchConfigMap.put(segmentNamePostfixProp, segmentSuffix);
 
+    // For upsert tables must use the UploadedRealtimeSegmentName for right assignment of segments
+    if (_tableConfig.isUpsertEnabled()) {
+      batchConfigMap.put(BatchConfigProperties.SEGMENT_NAME_GENERATOR_TYPE,
+          BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME);
+    }
+
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:154]: Not introduced in this PR: Since right assignment of segments is strictly required in upsert tables and not other non-upsert realtime tables, do you think SegmentNameGeneratorType.UPLOADED_UPSERT would have made more sense?

Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:154]: The segment name though more useful for upsert is actually needed for all realtime tables to differentiate between stream generated segments and uploaded segments.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 2 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:53]: No prefernece here, the segment naming convention implies uploaded realtime segment, so putting uploaded in prefix seems redundant. Let me know if you see any issues from minion operations perspective.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 1 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:67]: The idea is to have either the job start time in segment name or if not provided then current time when segment is flused. Can rename it if this is the naming convention followed for similar scenarios.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 4]
```diff
@@ -124,12 +132,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
...
     String segmentNamePostfixProp = String.format("%s.%s", BatchConfigProperties.SEGMENT_NAME_GENERATOR_PROP_PREFIX,
         BatchConfigProperties.SEGMENT_NAME_POSTFIX);
     String segmentSuffix = batchConfigMap.get(segmentNamePostfixProp);
     segmentSuffix = segmentSuffix == null ? String.valueOf(_indexOfSubtask) : segmentSuffix + "_" + _indexOfSubtask;
     batchConfigMap.put(segmentNamePostfixProp, segmentSuffix);
 
+    // For upsert tables must use the UploadedRealtimeSegmentName for right assignment of segments
+    if (_tableConfig.isUpsertEnabled()) {
+      batchConfigMap.put(BatchConfigProperties.SEGMENT_NAME_GENERATOR_TYPE,
+          BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME);
```
Comment (ankitsultana) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:152]: Just want to call out that this will break scenarios where the flink connector is running on a new version but the Pinot cluster is running on an older version.

But I guess this should be fine because flink upsert backfill was not working anyways? @rohityadav1993

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 1 above)
Comment (tibrewalpratik17) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:67]: Hmm the variable name makes sense in `FlinkSegmentWriter` but in this class it looks like out of context imo.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 3 above)
Comment (tibrewalpratik17) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:154]: Hmm in that case should we check if the table is of realtime nature rather than only upsert table?

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 4 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:152]: Yes, this wasn't working for upsert tables as there was no segment assignment logic in controller for uploaded segments.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 5]
```diff
@@ -124,12 +132,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
         "batchConfigMaps must contain only 1 BatchConfig for table: %s", _tableNameWithType);
 
     Map<String, String> batchConfigMap = _batchIngestionConfig.getBatchConfigMaps().get(0);
+    batchConfigMap.put(BatchConfigProperties.PARTITION_ID, Integer.toString(_indexOfSubtask));
+    batchConfigMap.put(BatchConfigProperties.SEQUENCE_ID, Integer.toString(_seqId));
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:136]: Will this not always put the sequence-id as 0 in all scenarios? Can this cause conflicts when creation time is same as upsert table might end up replacing the old segment?

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 3 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:154]: We ideally should have it. I wouldn't be sure if somebody has made this work for non upsert table with LLCSegmentName or another name format with a minor patch in server side and with default segment assignment strategy, in that case this change will break uploads.

So was the reason to mandate it for upserts only which were broken until https://github.com/apache/pinot/pull/13107

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 1 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:67]: Let me rename it here with a code comment.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 3 above)
Comment (tibrewalpratik17) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:154]: In that case, they will have an internal implementation of `SegmentWriter` I guess. Anyways, if there are internal patches if people are using to make this working then better to have it in Pinot as first-class itself.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 5 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:136]: Thanks for catching this, it needs to be in flush as FlinkSegmentWriter gets initialised once. I'll add an integration test around this as well.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 3 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java]: Making `UploadedRealtimeSegmentName` as the only segment naming format supported to upload to a reltime table.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 5 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java]: Fixed. Adding integration test caught another pre-existing bug with flush happening at flushThreshold+1. Fixed as part of the same PR.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 6]
```diff
@@ -60,13 +63,19 @@ public class PinotSinkFunction<T> extends RichSinkFunction<T> implements Checkpo
...
+  @Nullable private Long _segmentUploadTimeMs;
+
   private transient SegmentWriter _segmentWriter;
   private transient SegmentUploader _segmentUploader;
   private transient ExecutorService _executor;
   private transient long _segmentNumRecord;
 
   public PinotSinkFunction(PinotGenericRowConverter<T> recordConverter, TableConfig tableConfig, Schema schema) {
     this(recordConverter, tableConfig, schema, DEFAULT_SEGMENT_FLUSH_MAX_NUM_RECORDS, DEFAULT_EXECUTOR_POOL_SIZE);
+    _segmentNamePrefix = DEFAULT_UPLOADED_REALTIME_SEGMENT_PREFIX;
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:78]: nit: this is redundant as the below ctor is already initialising  `_segmentNamePrefix`

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfigProperties.java [hunk 7]
```diff
@@ -43,7 +43,9 @@ private BatchConfigProperties() {
   public static final String TABLE_CONFIGS_URI = "tableConfigsURI";
   public static final String SCHEMA = "schema";
   public static final String SCHEMA_URI = "schemaURI";
+  public static final String PARTITION_ID = "partitionId";
   public static final String SEQUENCE_ID = "sequenceId";
+  public static final String SEGMENT_CREATION_TIME_MS = "segmentCreationTimeMs";
```
Comment (tibrewalpratik17) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfigProperties.java:48]: nit: we should name this as `segmentUploadTimeMs` too for consistency across code.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 8]
```diff
@@ -60,6 +63,11 @@ public class PinotSinkFunction<T> extends RichSinkFunction<T> implements Checkpo
   private TableConfig _tableConfig;
   private Schema _schema;
 
+  private String _segmentNamePrefix;
+
+  // Used to set creation time in segment name, if not provided, current time is used
```
Comment (tibrewalpratik17) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:68]:
```suggestion
  // Used to set upload time in segment name, if not provided, current time is used
```

**Review by tibrewalpratik17 — APPROVED** [review]
Just one nit rest LGTM!

**Review by tibrewalpratik17 — APPROVED** [review]

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java [hunk 9]
```diff
@@ -52,7 +52,9 @@ public class BatchConfig {
   private final boolean _excludeSequenceId;
   private final boolean _appendUUIDToSegmentName;
   private final boolean _excludeTimeInSegmentName;
+  private final String _partitionId;
```
Comment (ankitsultana) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java:55]: partition is quite overloaded.. this is the "uploadedRealtimePartitionId"?

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java [hunk 10]
```diff
@@ -185,6 +189,14 @@ public boolean isExcludeSequenceId() {
     return _excludeSequenceId;
   }
 
+  public String getPartitionId() {
+    return _partitionId;
+  }
+
+  public long getSegmentUploadTimeMs() {
```
Comment (ankitsultana) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java:196]: nit: getter order should match declaration order: _partitionId > sequenceId > segmentUploadTimeMs

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java [hunk 11]
```diff
@@ -169,6 +170,9 @@ private static SegmentNameGenerator getSegmentNameGenerator(BatchConfig batchCon
       case BatchConfigProperties.SegmentNameGeneratorType.SIMPLE:
         return new SimpleSegmentNameGenerator(rawTableName, batchConfig.getSegmentNamePostfix(),
             batchConfig.isAppendUUIDToSegmentName(), batchConfig.isExcludeTimeInSegmentName());
+      case BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME:
+        return new UploadedRealtimeSegmentNameGenerator(rawTableName, Integer.parseInt(batchConfig.getPartitionId()),
```
Comment (ankitsultana) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java:174]: Can we throw a graceful error when integer parse fails? The current error message will not be that intuitive and users will have to look at the code to debug.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 12]
```diff
@@ -119,7 +139,7 @@ public void invoke(T value, Context context)
       throws Exception {
     _segmentWriter.collect(_recordConverter.convertToRow(value));
     _segmentNumRecord++;
-    if (_segmentNumRecord > _segmentFlushMaxNumRecords) {
+    if (_segmentNumRecord >= _segmentFlushMaxNumRecords) {
```
Comment (ankitsultana) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:142]: curious, why is this required?

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 13]
```diff
@@ -76,13 +84,25 @@ public PinotSinkFunction(PinotGenericRowConverter<T> recordConverter, TableConfi
...
     _segmentFlushMaxNumRecords = segmentFlushMaxNumRecords;
     _executorPoolSize = executorPoolSize;
+    _segmentNamePrefix = DEFAULT_UPLOADED_REALTIME_SEGMENT_PREFIX;
+  }
+
+  public PinotSinkFunction(PinotGenericRowConverter<T> recordConverter, TableConfig tableConfig, Schema schema,
+      long segmentFlushMaxNumRecords, int executorPoolSize, String segmentNamePrefix,
+      @Nullable Long segmentUploadTimeMs) {
+    this(recordConverter, tableConfig, schema, segmentFlushMaxNumRecords, executorPoolSize);
+    if (!segmentNamePrefix.isBlank()) {
```
Comment (ankitsultana) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:94]: this means we enforce a non-empty segment name prefix? could there be a need to support no prefix?

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 14]
```diff
@@ -95,8 +100,12 @@ public class FlinkSegmentWriter implements SegmentWriter {
   private transient Counter _processedRecords;
   private transient volatile long _lastRecordProcessingTimeMs = 0;
 
-  public FlinkSegmentWriter(int indexOfSubtask, MetricGroup metricGroup) {
+  public FlinkSegmentWriter(int indexOfSubtask, MetricGroup metricGroup, String segmentNamePrefix,
```
Comment (ankitsultana) [inline pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:103]: b/w incompatible change is not a concern here? I guess callers only call the sink function?

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java (hunk 9 above)
Comment (rohityadav1993) [reply pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java:55]: agree, this is more clear.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 12 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:142]: This was a bugfix to make flush happen at `_segmentFlushMaxNumRecords` rows and not `_segmentFlushMaxNumRecords + 1`

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 14 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:103]: right, PinotSinkFunction is exposed for Flink and not a whole lot of benefit in using FlinkSegmentWriter directly.
There is a `FileBasedSegmentWriter` implementing the interface `SegmentWriter` for general use.

Still, let me call this out in the description and mark backward incompatible.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java (hunk 13 above)
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:94]: It would depend on the segmentNameGenerator property if the prefix is needed, even though configured. But rethinking this, there are two ways to configure this:
- In table config as batchIngestionConfig
- In code via PinotSinkFunction config -> SegmentGenerator config

Right now, latter will always take precedence with a default value. Refactoring so that default value is used when neither is configured otherwise

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java [hunk 15]
```diff
@@ -169,6 +170,16 @@ private static SegmentNameGenerator getSegmentNameGenerator(BatchConfig batchCon
       case BatchConfigProperties.SegmentNameGeneratorType.SIMPLE:
         return new SimpleSegmentNameGenerator(rawTableName, batchConfig.getSegmentNamePostfix(),
             batchConfig.isAppendUUIDToSegmentName(), batchConfig.isExcludeTimeInSegmentName());
+      case BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME:
+        int uploadedRealtimePartitionId;
+        try {
+          uploadedRealtimePartitionId = Integer.parseInt(batchConfig.getUploadedRealtimePartitionId());
+        } catch (NumberFormatException e) {
+          throw new IllegalStateException(
```
Comment (ankitsultana) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java:178]: nit: `IllegalArgumentException`?

**Review by ankitsultana — APPROVED** [review]

**Review by Jackie-Jiang — COMMENTED** [review]
Mostly good

**Comment by Jackie-Jiang** [comment]
cc @swaminathanmanish

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 16]
```diff
@@ -124,12 +135,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
         "batchConfigMaps must contain only 1 BatchConfig for table: %s", _tableNameWithType);
 
     Map<String, String> batchConfigMap = _batchIngestionConfig.getBatchConfigMaps().get(0);
+    batchConfigMap.put(BatchConfigProperties.UPLOADED_REALTIME_PARTITION_ID, Integer.toString(_indexOfSubtask));
```
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:138]: The partitioning is controlled by Flink partioning logic (see README). This was part of the original implementation, where `RichFunction.getRuntimeContext().getIndexOfThisSubtask()` provides the partition number if Flink parallelism is the same as the number of partitions. Same is being used to set the partition id for segment name generation.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java [hunk 17]
```diff
@@ -124,12 +135,26 @@ public void init(TableConfig tableConfig, Schema schema, Map<String, String> bat
...
+    batchConfigMap.put(BatchConfigProperties.SEGMENT_UPLOAD_TIME_MS, String.valueOf(_segmentUploadTimeMs));
+    batchConfigMap.computeIfAbsent(
+        BatchConfigProperties.SEGMENT_NAME_GENERATOR_PROP_PREFIX + "." + BatchConfigProperties.SEGMENT_NAME_PREFIX,
```
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:141]: Did not understand the behaviour change. `flink` is supposed to be a default prefix as `UploadedRealtimeSegmentName` mandates a prefix, one can use the batch config to modify the default value and handling of `UploadedRealtimeSegmentName` in Pinot is agnostic of the prefix keyword.

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java [hunk 18]
```diff
@@ -52,7 +52,9 @@ public class BatchConfig {
   private final boolean _excludeSequenceId;
   private final boolean _appendUUIDToSegmentName;
   private final boolean _excludeTimeInSegmentName;
+  private final String _uploadedRealtimePartitionId;
```
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfig.java:55]: Would want to handle cases where _uploadedRealtimePartitionId is not set but needed. Can use Integer for clarity of type.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java [hunk 19]
```diff
@@ -78,11 +84,20 @@ public PinotSinkFunction(PinotGenericRowConverter<T> recordConverter, TableConfi
     _executorPoolSize = executorPoolSize;
   }
 
+  public PinotSinkFunction(PinotGenericRowConverter<T> recordConverter, TableConfig tableConfig, Schema schema,
+      long segmentFlushMaxNumRecords, int executorPoolSize, String segmentNamePrefix,
```
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/PinotSinkFunction.java:88]: Nullable, thanks for the suggestion.

File: pinot-connectors/pinot-flink-connector/README.md [hunk 20]
```diff
@@ -39,9 +44,34 @@ srcDs.addSink(new PinotSinkFunction<>(new PinotRowRecordConverter(TEST_TYPE_INFO
...
+
+// Create a ControllerRequestClient to fetch Pinot schema and table config
+HttpClient httpClient = HttpClient.getInstance();
+ControllerRequestClient client = new ControllerRequestClient(
+ControllerRequestURLBuilder.baseUrl(DEFAULT_CONTROLLER_URL), httpClient);
+
+// fetch Pinot schema
+Schema schema = PinotConnectionUtils.getSchema(client, "starbucksStores");
+// fetch Pinot table config
+TableConfig tableConfig = PinotConnectionUtils.getTableConfig(client, "starbucksStores", "OFFLINE");
```
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/README.md:62]: realtime

**Comment by kishoreg** [comment]
Great addition. I had a few high level question
- is relying on the naming convention of the segment new or existing requirement. I am guessing it's existing. Is it possible to get rid of this requirement and make it part of the segment metadata? We can do this as part of a separate PR.
- can this be extended to having a batch(Offline) only upsert table?

**Comment by rohityadav1993** [comment]
@kishoreg , extracting partitionId from segmentName is a reuse of [LLCSegmentName](https://github.com/apache/pinot/blob/master/pinot-common/src/main/java/org/apache/pinot/common/utils/LLCSegmentName.java) convention in UploadedReatimeSegmentName introduced in PR #13107
A call was taken to utilise segmentName encoding to reduce overhead of querying Zk.

Yes, we can support batch-only upsert table. I had a few top of my thoughts:
 - Upsert table is essentially a realtime table and canont be created without a stream config.
 - Offline upsert tables would the right classification though behind the scenes it can be a realtime table with disabled stream consumption.
 - Many batch workloads have the possibility to deduplicate data during offline ingestion so offline tables are still usable if we overwrite all the data. Real advantage will be for usecases with incremental data ingestion with upsert requirement.

I anticipate a few usecases coming up in future for us and will be happy to enhance batch ingestion for upserts.

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (Jackie-Jiang) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:145]: I believe this writer is shared for both upsert and non-upsert table. It will be behavior change if we add `flink` prefix to non-upsert table

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 4 above)
Comment (Jackie-Jiang) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:159]: Consider adding a check for upsert table, and add checks under that if branch

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:145]: Makes sense, mandating UploadedRealtimeSegmentName naming can be only kept for upsert table which didn't work before. Will update.

Though, I do think we can make UploadedRealtimeSegmentName as default naming convention for uploaded segments for non upsert as well as it will lead to a better segment assignment which currently happens by taking mod with 10_000


```
    if (segmentPartitionId == null) {
      // This case is for the uploaded segments for which there's no partition information.
      // A random, but consistent, partition id is calculated based on the hash code of the segment name.
      // Note that '% 10K' is used to prevent having partition ids with large value which will be problematic later in
      // instance assignment formula.
      segmentPartitionId = Math.abs(segmentName.hashCode() % 10_000);
  }
```

File: pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfigProperties.java [hunk 21]
```diff
@@ -43,7 +43,9 @@ private BatchConfigProperties() {
   public static final String TABLE_CONFIGS_URI = "tableConfigsURI";
   public static final String SCHEMA = "schema";
   public static final String SCHEMA_URI = "schemaURI";
+  public static final String UPLOADED_REALTIME_PARTITION_ID = "uploadedRealtimePartitionId";
```
Comment (klsince) [inline pinot-spi/src/main/java/org/apache/pinot/spi/ingestion/batch/BatchConfigProperties.java:46]: nit: just call it `segmentPartitionId`?

per the naming convention here, perhaps better to name them like below:
```
segment.partitionId
segment.uploadTimeMs
```

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java [hunk 22]
```diff
@@ -169,6 +170,16 @@ private static SegmentNameGenerator getSegmentNameGenerator(BatchConfig batchCon
...
+      case BatchConfigProperties.SegmentNameGeneratorType.UPLOADED_REALTIME:
+        int uploadedRealtimePartitionId;
+        try {
+          uploadedRealtimePartitionId = Integer.parseInt(batchConfig.getUploadedRealtimePartitionId());
+        } catch (NumberFormatException e) {
+          throw new IllegalArgumentException(
+              String.format("Invalid uploadedRealtimePartitionId: %s", batchConfig.getUploadedRealtimePartitionId()));
+        }
+        return new UploadedRealtimeSegmentNameGenerator(rawTableName, uploadedRealtimePartitionId,
+            batchConfig.getSegmentUploadTimeMs(), batchConfig.getSegmentNamePrefix(), batchConfig.getSequenceId());
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java:182]: how many segments are generated by one flink task? I wonder if the sequentId should be adjusted to avoid overwriting segments uploaded by one task.

Comment (rohityadav1993) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java:182]: The segment count can be unbounded. The sequenceId gets generated per task i.e. if parallelism is 4, there will be 4 tasks each with their FlinkSegmentWriter instance and sequenceId gets incremented after each segment flush and upload.

Given sequenceId is incremental and partitionId differentiates each segment name, there isn't a possibility of overwrite by another task's same sequenceId.

Comment (klsince) [reply pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/IngestionUtils.java:182]: I see, and I just spotted this in the README that each task will generate segments for one particular table partition, then it make sense to use a sequenceId local to each task.
```
The parallelism of the job *must* be set the same as the number of partitions of the Pinot table(same as upstream), so that the sink in each task executor can generate the segment of same partitions.
```

**Review by klsince — APPROVED** [review]

File: pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (rohityadav1993) [reply pinot-connectors/pinot-flink-connector/src/main/java/org/apache/pinot/connector/flink/sink/FlinkSegmentWriter.java:144]: Refactored to use the naming convention only for upsert table:
https://github.com/apache/pinot/pull/13837/files#diff-72b89e65ea292a808b5d8a902d8a64adb87834534414d5b9d0e46c4725a0455cR156
